### ISSUE TITLE
v0.5.0 alpha randomly stops syncing

### ISSUE THREAD

**Issue by Avigant** [LPGhatguy/rojo#110]
I have absolutely no idea how to reproduce this, but it is happening once or twice an hour. Serving never stops, only syncing. The plugin shows that Rojo is no longer connected. I am on Windows 10.

**Comment by Avigant** [comment]
Syncing works as soon as I hit the "connect" button on the plugin again. Nothing happens in my Output. I should've specified this.

I will visit that at the next time I encounter this and report back.

**Comment by LPGhatguy** [comment]
Okay, that issue in the output window helps a lot, I should be able to track that down this week!

**Comment by Validark** [comment]
I encountered this issue too. For me it happens after I stop PlaySolo.

**Comment by Validark** [comment]
When it happens, it will quickly call `ApiContext:retrieveMessages()`'s `andThen` method twice.

The first one seems normal
```
{body = "{"sessionId":"9bf75c92-4e56-499d-8242-8e2b5511cc37","messageCursor":4,"messages":[{"added":[],"removed":[],"updated":["91f67966-8457-4c89-8a0b-63cb9be819da"]}]}", code = 200, headers = {["content-length"] = "160", ["content-type"] = "application/json", date = "Tue, 22 Jan 2019 21:06:11 GMT", server = "tiny-http (Rust)"}} <- {__index = [self], __tostring = function, fromRobloxResponse = function, isSuccess = function, json = function}
```
Then afterwards, it will call it with a table that looks like this:
```
{{added = {}, removed = {}, updated = {"91f67966-8457-4c89-8a0b-63cb9be819da"}}}
```

**Comment by LPGhatguy** [comment]
I refactored a lot of the relevant code in the plugin recently, can you see if the problem is still happening in the latest alpha?

**Comment by Validark** [comment]
I am curious how that random table somehow leaked into the Promise. How was it even possible? Have any theories?

**Comment by LPGhatguy** [comment]
I figured it out specifically today and fixed it. It was pretty simple!

In `ApiContext:retrieveMessages()`, I was trying to respond to a timeout by restarting listening for messages again:

https://github.com/LPGhatguy/rojo/blob/4516df5aac4c9296ba30256d5437e9ffe06564f9/plugin/src/ApiContext.lua#L140-L163

Note that on line 146, I'm calling `retrieveMessages` again. `retrieveMessages` doesn't return an `HttpResponse`, it returns a table that looks like the second one you reported!

Basically, if there was a timeout (and thus another `retrieveMessages` call) and that call _did not_ time out, those changes would be applied followed immediately by an error.

The fix was straightforward. I made a new function to do the recursive timeout nonsense and I just invoke it from within `retrieveMessages` now:

https://github.com/LPGhatguy/rojo/blob/aa243d1b8af49731e9977dd6f0413652f0b5dd07/plugin/src/ApiContext.lua#L140-L167
